File: app/src/androidTest/java/com/better/alarm/test/PopupTest.java
package com.better.alarm.test;

import static android.provider.AlarmClock.ACTION_SET_ALARM;
import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.longClick;
import static androidx.test.espresso.matcher.ViewMatchers.isChecked;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.assertj.core.api.Assertions.assertThat;
import static org.hamcrest.Matchers.allOf;

import android.content.Intent;
import android.provider.AlarmClock;
import androidx.test.rule.ActivityTestRule;
import androidx.test.runner.AndroidJUnit4;
import com.better.alarm.R;
import com.better.alarm.alert.AlarmAlertFullScreen;
import com.better.alarm.configuration.ContainerKt;
import com.better.alarm.configuration.InjectKt;
import com.better.alarm.configuration.Store;
import com.better.alarm.interfaces.Intents;
import com.better.alarm.model.AlarmSetter;
import com.better.alarm.model.AlarmValue;
import com.better.alarm.model.AlarmsReceiver;
import com.better.alarm.model.CalendarType;
import com.better.alarm.presenter.AlarmsListActivity;
import com.better.alarm.presenter.HandleSetAlarm;
import com.better.alarm.presenter.TransparentActivity;
import com.better.alarm.util.Optional;
import java.util.Calendar;
import java.util.Locale;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.RuleChain;
import org.junit.rules.TestRule;
import org.junit.runner.RunWith;

/** Created by Yuriy on 12.07.2017. */
@RunWith(AndroidJUnit4.class)
public class PopupTest extends BaseTest {
  public ActivityTestRule<AlarmAlertFullScreen> alertActivity =
      new ActivityTestRule<AlarmAlertFullScreen>(
          AlarmAlertFullScreen.class, false, /* autostart*/ false);
  public ActivityTestRule<AlarmsListActivity> listActivity =
      new ActivityTestRule<AlarmsListActivity>(
          AlarmsListActivity.class, false, /* autostart*/ true);
  public ActivityTestRule<TransparentActivity> transparentActivity =
      new ActivityTestRule<TransparentActivity>(
          TransparentActivity.class, false, /* autostart*/ false);

  @Rule
  public TestRule chain =
      RuleChain.outerRule(new ForceLocaleRule(Locale.US))
          .around(listActivity)
          .around(alertActivity)
          .around(transparentActivity);

  @Before
  public void set24HourMode() {
    ContainerKt.overrideIs24hoursFormatOverride(true);
  }

  public int createAlarmAndFire() {
    Intent createAlarm = new Intent();
    createAlarm.setClass(listActivity.getActivity(), HandleSetAlarm.class);
    createAlarm.setAction(ACTION_SET_ALARM);
    createAlarm.putExtra(AlarmClock.EXTRA_SKIP_UI, true);
    createAlarm.putExtra(AlarmClock.EXTRA_HOUR, 0);
    createAlarm.putExtra(AlarmClock.EXTRA_MINUTES, 0);
    createAlarm.putExtra(AlarmClock.EXTRA_MESSAGE, "From outside");

    listActivity.getActivity().startActivity(createAlarm);

    sleep();

    assertThatList().filter(enabled()).items().hasSize(1);

    int id =
        ListAsserts.<AlarmValue>listObservable(R.id.list_fragment_list)
            .firstOrError()
            .blockingGet()
            .getId();

    // simulate alarm fired
    Intent intent = new Intent(AlarmSetter.ACTION_FIRED);
    intent.setClass(listActivity.getActivity(), AlarmsReceiver.class);
    intent.putExtra(AlarmSetter.EXTRA_ID, id);
    intent.putExtra(AlarmSetter.EXTRA_TYPE, CalendarType.NORMAL.name());

    listActivity.getActivity().sendBroadcast(intent);

    sleep();
    return id;
  }

  @Test
  public void dismissViaAlarmAlert() {
    int id = createAlarmAndFire();

    Intent startIntent = new Intent();
    startIntent.putExtra(Intents.EXTRA_ID, id);
    alertActivity.launchActivity(startIntent);

    sleep();

    onView(withText("Dismiss")).perform(longClick());

    deleteAlarm();
  }

  private void afterLongClickSnoozeAlarmCheckAndDelete() {
    assertTimerView("--:--");
    onView(withText("2")).perform(click());
    assertTimerView("--:-2");
    onView(withText("3")).perform(click());
    assertTimerView("--:23");
    onView(withText("5")).perform(click());
    assertTimerView("-2:35");
    onView(withText("9")).perform(click());
    assertTimerView("23:59");
    sleep();

    onView(withText("OK")).perform(click());

    sleep();

    Optional<Store.Next> next =
        InjectKt.globalInject(Store.class).getValue().next().blockingFirst();

    assertThat(next.isPresent()).isTrue();

    Calendar nextTime = Calendar.getInstance();
    nextTime.setTimeInMillis(next.get().nextNonPrealarmTime());

    assertThat((int) nextTime.get(Calendar.HOUR_OF_DAY)).isEqualTo(23);
    assertThat((int) nextTime.get(Calendar.MINUTE)).isEqualTo(59);
    assertThat(next.get().alarm().isEnabled()).isTrue();

    // disable the snoozed alarm
    onView(allOf(withId(R.id.list_row_on_off_switch), isChecked())).perform(click());

    sleep();

    assertThatList().filter(enabled()).items().isEmpty();

    deleteAlarm();
  }

  @Ignore("Flakey")
  @Test
  public void snoozeViaClick() {
    int id = createAlarmAndFire();

    Intent startIntent = new Intent();
    startIntent.putExtra(Intents.EXTRA_ID, id);
    alertActivity.launchActivity(startIntent);

    sleep();
    onView(withText("Snooze")).perform(click());

    deleteAlarm();
  }

  @Ignore
  @Test
  public void snoozeViaNotificationPicker() {
    int id = createAlarmAndFire();

    Intent startIntent = new Intent();
    startIntent.putExtra(Intents.EXTRA_ID, id);
    transparentActivity.launchActivity(startIntent);

    deleteAlarm();
  }

  @Test
  public void letAlarmExpireAndDismissIt() throws InterruptedException {
    int id = createAlarmAndFire();

    Intent startIntent = new Intent();
    startIntent.putExtra(Intents.EXTRA_ID, id);
    alertActivity.launchActivity(startIntent);

    // simulate timed out
    Intent intent = new Intent(AlarmSetter.ACTION_FIRED);
    intent.setClass(alertActivity.getActivity(), AlarmsReceiver.class);
    intent.putExtra(AlarmSetter.EXTRA_ID, id);
    intent.putExtra(AlarmSetter.EXTRA_TYPE, CalendarType.AUTOSILENCE.name());

    listActivity.getActivity().sendBroadcast(intent);
    sleep();

    // Popup must be closed in order for this to work
    Thread.sleep(1000);
    deleteAlarm();
  }

  private void deleteAlarm() {
    sleep();
    deleteAlarm(0);
    assertThatList().items().hasSize(2);
  }
}

File: app/src/androidTest/java/com/better/alarm/test/PopupTest.kt
package com.better.alarm.test

import android.app.Activity
import android.content.Intent
import android.provider.AlarmClock
import androidx.lifecycle.Lifecycle
import androidx.test.core.app.ActivityScenario
import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.ext.junit.rules.ActivityScenarioRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.better.alarm.R
import com.better.alarm.bootstrap.overrideIs24hoursFormatOverride
import com.better.alarm.data.CalendarType
import com.better.alarm.domain.AlarmSetter
import com.better.alarm.domain.Store
import com.better.alarm.receivers.AlarmsReceiver
import com.better.alarm.receivers.Intents
import com.better.alarm.ui.alert.AlarmAlertFullScreen
import com.better.alarm.ui.alert.TransparentActivity
import com.better.alarm.ui.exported.HandleSetAlarm
import com.better.alarm.ui.main.AlarmsListActivity
import java.util.*
import java.util.stream.Collectors
import org.assertj.core.api.Assertions.assertThat
import org.hamcrest.Matchers
import org.junit.After
import org.junit.Before
import org.junit.Ignore
import org.junit.Rule
import org.junit.Test
import org.junit.rules.RuleChain
import org.junit.rules.TestRule
import org.junit.runner.RunWith
import org.koin.core.context.GlobalContext

/** Created by Yuriy on 12.07.2017. */
@RunWith(AndroidJUnit4::class)
class PopupTest {

  var listActivity = ActivityScenarioRule(AlarmsListActivity::class.java)

  @JvmField
  @Rule
  var chain: TestRule =
      RuleChain.outerRule(ForceLocaleRule(Locale.US)).around(listActivity) // .around(alertActivity)

  @Before
  fun before() {
    overrideIs24hoursFormatOverride(true)
    dropDatabase()
  }

  @After
  @Throws(InterruptedException::class)
  fun tearDown() {
    dropDatabase()
  }

  private fun createAlarmAndFire(): Int {
    val context = InstrumentationRegistry.getInstrumentation().targetContext
    context.startActivity(
        Intent().apply {
          setClass(context, HandleSetAlarm::class.java)
          addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
          action = AlarmClock.ACTION_SET_ALARM
          putExtra(AlarmClock.EXTRA_SKIP_UI, true)
          putExtra(AlarmClock.EXTRA_HOUR, 0)
          putExtra(AlarmClock.EXTRA_MINUTES, 0)
          putExtra(AlarmClock.EXTRA_MESSAGE, "From outside")
        })
    assertThat(alarmsList().filter { it.isEnabled }).hasSize(1)
    val id = alarmsList()[0].id

    // simulate alarm fired
    context.sendBroadcast(
        Intent(AlarmSetter.ACTION_FIRED).apply {
          setClass(context, AlarmsReceiver::class.java)
          putExtra(AlarmSetter.EXTRA_ID, id)
          putExtra(AlarmSetter.EXTRA_TYPE, CalendarType.NORMAL.name)
        })
    return id
  }

  @Test
  fun dismissViaAlarmAlert() {
    val id = createAlarmAndFire()
    launchActivity(AlarmAlertFullScreen::class.java, id)
    onView().with(text = "Dismiss").perform(ViewActions.longClick())

    launchActivity(AlarmsListActivity::class.java, id)
    // delete after dismiss
    assertThat(alarmsList()).hasSize(2)
    // all disabled
    assertThat(alarmsList().filter { it.isEnabled }).hasSize(0)
  }

  private fun launchActivity(activityClass: Class<out Activity>, id: Int) {
    val context = InstrumentationRegistry.getInstrumentation().targetContext
    ActivityScenario.launch<Activity>(
            Intent(context, activityClass).apply { putExtra(Intents.EXTRA_ID, id) })
        .moveToState(Lifecycle.State.RESUMED)
  }

  private fun afterLongClickSnoozeAlarmCheckAndDelete() {
    assertTimerView("--:--")
    Espresso.onView(ViewMatchers.withText("2")).perform(ViewActions.click())
    assertTimerView("--:-2")
    Espresso.onView(ViewMatchers.withText("3")).perform(ViewActions.click())
    assertTimerView("--:23")
    Espresso.onView(ViewMatchers.withText("5")).perform(ViewActions.click())
    assertTimerView("-2:35")
    Espresso.onView(ViewMatchers.withText("9")).perform(ViewActions.click())
    assertTimerView("23:59")
    Espresso.onView(ViewMatchers.withText("OK")).perform(ViewActions.click())
    val next = GlobalContext.get().get<Store>().next().blockingFirst()
    assertThat(next.isPresent()).isTrue
    val nextTime = Calendar.getInstance()
    nextTime.timeInMillis = next.get().nextNonPrealarmTime()
    assertThat(nextTime[Calendar.HOUR_OF_DAY]).isEqualTo(23)
    assertThat(nextTime[Calendar.MINUTE]).isEqualTo(59)
    assertThat(next.get().alarm().isEnabled).isTrue

    // disable the snoozed alarm
    Espresso.onView(
            Matchers.allOf(
                ViewMatchers.withId(R.id.list_row_on_off_switch), ViewMatchers.isChecked()))
        .perform(ViewActions.click())
    assertThat(alarmsList().stream().filter(enabled()).collect(Collectors.toList())).isEmpty()
  }

  @Ignore("Flaky")
  @Test
  fun snoozeViaClick() {
    val id = createAlarmAndFire()
    launchActivity(AlarmAlertFullScreen::class.java, id)
    onView().with(text = "Snooze").click()
  }

  @Ignore("Flaky")
  @Test
  fun snoozeViaNotificationPicker() {
    val id = createAlarmAndFire()
    launchActivity(TransparentActivity::class.java, id)
    afterLongClickSnoozeAlarmCheckAndDelete()
  }

  @Test
  fun letAlarmExpireAndDismissIt() {
    val id = createAlarmAndFire()
    launchActivity(AlarmAlertFullScreen::class.java, id)

    // simulate timed out
    val context = InstrumentationRegistry.getInstrumentation().targetContext
    context.sendBroadcast(
        Intent(context, AlarmsReceiver::class.java).apply {
          action = AlarmSetter.ACTION_FIRED
          putExtra(AlarmSetter.EXTRA_ID, id)
          putExtra(AlarmSetter.EXTRA_TYPE, CalendarType.AUTOSILENCE.name)
        })

    launchActivity(AlarmsListActivity::class.java, id)
    // delete after dismiss
    assertThat(alarmsList()).hasSize(2)
    assertThat(alarmsList().filter { it.isEnabled }).hasSize(0)
  }
}
